nncf/experimental/common/pruning/nodes_grouping.py

```diff
@@ -0,0 +1,466 @@
+                graph.add_edge(parent_graph_id, child_graph_id)
+                add_group_to_graph(graph, child_group, visited_block_ids_map)
+            else:
+                child_graph_id = visited_block_ids_map[child_id]
+                graph.add_edge(parent_graph_id, child_graph_id)
+
+
+counter = 0
+
+
+def build_block_hierarchy(root_groups: List[PropagationGroup]) -> nx.DiGraph:
+    """Creates hierarchy of propagation blocks/groups by traversing children in the given root groups.
+
+    :param roots: list of the root groups
+    :return: networkx graph that represents the hierarchy of propagation blocks/groups.
+    """
+    graph = nx.DiGraph()
+    global counter
+    counter = 0
+    visited = dict()
+    for root_group in root_groups:
+        add_group_to_graph(graph, root_group, visited)
+        counter += 1
+    return graph
+
+
+def get_pruning_groups(graph: NNCFGraph,
+                       pruning_operations_metatypes: PruningOperationsMetatypeRegistry,
+                       prune_operations_types: List[str],
+                       dump_dir: Optional[Path] = None) -> List[PruningGroup]:
+    """
+    Determines how nodes of the given types should be pruned: which nodes should be pruned together, along which
+    dimension, how many sequent channels with which offset. It's done by initializing PropagationMask's on the
+    operations with prunable paramaters (producers of pruning) and propagating them through the execution graph.
+
+    :param graph: nncf graph to initialize and propagate masks.
+    :param pruning_operations_metatypes: registry with operation metatypes pruning algorithm is aware of, i.e.
+    metatypes describing operations with common pruning mask application and propagation properties, e.g.
+    IdentityMaskForwardOps unifies operations that propagate pruning masks as is (relu, swish etc.), whereas
+    Convolution unifies different convolution operations (conv1d, conv2d, conv3d) which accepts some input masks and
+    provide some output masks.
+    :param prune_operations_types: types of operations with prunable parameters.
+    :param dump_dir: path to the directory for dumping debug files.
+    :return: list of groups with parameters of pruning.
+    """
+    # 1. Initialize masks for producing nodes
+    all_nodes_to_prune = graph.get_nodes_by_types(prune_operations_types)  # type: List[NNCFNode]
+    roots = {}
+    for node in all_nodes_to_prune:
+        assert isinstance(node.layer_attributes, (LinearLayerAttributes, ConvolutionLayerAttributes))
+        pruning_dim = node.layer_attributes.get_target_dim_for_compression()
+        output_tensors_shapes = [x.tensor_shape for x in graph.get_output_edges(node)]
+        assert len(output_tensors_shapes) == 1 or len( set(output_tensors_shapes)) <= 1, node.node_name
+        output_tensors_shape = output_tensors_shapes[0]
+        target_output_dim_for_compression = len(output_tensors_shape) - 1
```

Comment:
So, compression dim for linear and conv layers is always last dim?

Reply:
For linear – yes, pruning convolutions is out of scope now. Ideally, there should be a method  `get_output_dim_affected_by_compression` in the layer attributes. Potentially, there could be a difference between frameworks, couldn't it?

Reply:
Yes, there are differences even between TF and Torch. Could you please add todo and update a ticket with this finding?